### PR TITLE
Automatically generate literal models via codegen

### LINKED ISSUE #63: Runtime created classes do not afford user-friendly interface
The way that classes are created during runtime does not afford auto-completion.
There are a few ways to approach this problem.

One would be to overload a few dunder methods to allow IDEs to list object attributes. However this doesn't fix the problem since, before the object is created, the attributes do not exist.

Another way to approach this problem is to take inspiration from automatically generated C wrappers and simply create all classes via templating.

### PR DESCRIPTION
**Pull request by bruno-f-cruz**
This PR attempts to fix #63 by automatically generating models

I should preface this PR by saying that the underlying pattern of these enum-like classes is not ideal and smells a bit like an anti-pattern to me. Mainly because of https://github.com/AllenNeuralDynamics/aind-data-schema/issues/960

Several of the patterns I used for the generators reflect the following discussions:
  - Improving Union methods as described in https://github.com/AllenNeuralDynamics/aind-data-schema/issues/843 and https://github.com/AllenNeuralDynamics/aind-data-schema/discussions/794
  - Generating classes from an endpoint instead of CSV files (this could be a suitable solution for #59  and #51 )
  - Removing the CSV files from the pypi package to prevent bloating
  - Improve type hinting (https://github.com/AllenNeuralDynamics/aind-data-schema/issues/843)

### CODE REVIEW

**Comment by bruno-f-cruz** [comment]
I should add that this PR could also be extended by:
  - Generating classes from an endpoint instead of CSV files
  - Removing the CSV files from the pypi package to prevent bloating

**Comment by bruno-f-cruz** [comment]
I have added an example in https://github.com/AllenNeuralDynamics/aind-data-schema-models/pull/64/commits/bc3dc8b741df7f26c853f4ffa80fcf5603efc85f on how to derive models from a remote source. In this case, harp-tech, keeps a list of all harp devices. Since all harp devices must reserve a whoami at the harp-tech website, we can simply use that list to generate our models. In order to version the list, we can lock it to a specific commit. This way the list will always be compatible with registered devices, will also encourage people to follow the standard, and remove several sources of human labor and error.

This pattern could be easily extended to other remote sources.

**Comment by bruno-f-cruz** [comment]
The coverage settings may need to change in order for tests to pass. I will defer to a maintainer for this decision...

File: generate.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        parser=lambda: csv_parser(root / "modalities.csv"),
+    )
+
+    harp_device_types = ModelGenerator(
+        enum_like_class_name="_HarpDeviceType",
+        parent_model_type=_HarpDeviceTypeModel,
+        discriminator="name",
+        data_source_identifier="https://raw.githubusercontent.com/harp-tech/protocol/97ded281bd1d0d7537f90ebf545d74cf8ba8805e/whoami.yml",  # noqa: E501
+        parser=lambda: get_who_am_i_list(
+            url="https://raw.githubusercontent.com/harp-tech/protocol/97ded281bd1d0d7537f90ebf545d74cf8ba8805e/whoami.yml"  # noqa: E501
```
Comment (jtyoung84) [inline generate.py:46]: I wonder if we can separate this out into a different PR?

File: generate.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    GeneratorContext,
+    MappableReferenceField,
+    ModelGenerator,
+    ForwardClassReference,
+)
+from aind_data_schema_models.generators.parsers import csv_parser, get_who_am_i_list
+
+if __name__ == "__main__":
+
+    root = Path(__file__).parent / r"src/aind_data_schema_models/models"
```
Comment (jtyoung84) [inline generate.py:21]: I think this might not translate across different operating systems? Maybe
```python
root = Path(__file__).parent / "src" / "aind_data_schema_models" / "models"
```

File: generate.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    target_folder = Path(r".\src\aind_data_schema_models\_generated")
```
Comment (jtyoung84) [inline generate.py:22]: We could probably have this use a similar way that root is defined?

File: generate.py (hunk 1 above)
Comment (bruno-f-cruz) [reply generate.py:46]: Sounds good!

File: generate.py (hunk 2 above)
Comment (bruno-f-cruz) [reply generate.py:21]: It should. The Path class should gracefully handle different operating systems. Keep in mind that the "/" is an overload of the `__truediv__` dunder operator not a literal.

File: generate.py (hunk 3 above)
Comment (jtyoung84) [reply generate.py:22]: We could put the generator code in a separate repo and bundle it into a docker image (or pypi package). We could then just put the instructions in the readme like: 
`docker run -e {parameters specifying the input and output directories} name-of-docker-image:latest`

**Comment by dbirman** [comment]
I'd like to get this merged (but with the external registry removed). Do you have time to split this into separate PRs? If not, I can split the code into branches in the main models repo. Not sure if there's an easy way to do that and preserve your commit history though.

**Comment by bruno-f-cruz** [comment]
Closing in favor of #91
